walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -24,33 +24,28 @@
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    // TODO: Maybe use binary tree or another data structure for availableTimes 
-    // because branches split in two?
-    // TODO: Combine the repetition of calling getAvailableTimes() below so it
-    // doesn't need to repeat without optional attendees.
     Collection<TimeRange> availableTimes = new ArrayList<>();
+    Collection<TimeRange> availableTimesWithoutOptional;
     Collection<String> attendees = request.getAttendees();
     Collection<String> optionalAttendees = request.getOptionalAttendees();
-    Collection<String> allAttendees = new ArrayList<>(request.getAttendees());
-    allAttendees.addAll(optionalAttendees);
     long duration = request.getDuration();
     if (duration > TimeRange.WHOLE_DAY.duration()){
       return availableTimes;
     }
-    availableTimes = getAvailableTimes(events, allAttendees, duration);
-    if (availableTimes.isEmpty()){
-      availableTimes = getAvailableTimes(events, attendees, duration);
+    availableTimes.add(TimeRange.WHOLE_DAY);
+    getAvailableTimes(availableTimes, events, attendees, duration);
+    if (!availableTimes.isEmpty()){
+      return getAvailableTimesWithMaxOptional(availableTimes, events, optionalAttendees, duration);
+    } else {
+      return availableTimes;
     }
-    return availableTimes;
   }
 
-  private Collection<TimeRange> getAvailableTimes(Collection<Event> events, Collection<String> attendees, long duration){
-    Collection<TimeRange> availableTimes = new ArrayList<>();
-    Collection<TimeRange> pendingTimes = new ArrayList<>();
-    availableTimes.add(TimeRange.WHOLE_DAY);
+  private void getAvailableTimes(Collection<TimeRange> availableTimes, Collection<Event> events, Collection<String> attendees, long duration){
     if (attendees.isEmpty()){
-      return availableTimes;
+      return;
     }
+    Collection<TimeRange> pendingTimes = new ArrayList<>();
     for (Event event: events){
       for (String attendee: attendees){
```

Comment:
If we have 1000 events and 1000 attendees, and say event #998 books mandatory attendee #998 for 12am to 10pm.  How quick before the algorithm detects that only slots 10pm to 11:59pm are open?  Any way we can have it fail quicker for all the earlier time slots?

Reply:
Also if attendee #2 doesn't have any booked events.  How many times do we need to evaluate attendee #2.